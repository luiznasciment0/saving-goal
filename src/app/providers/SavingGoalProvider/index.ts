import constate from 'constate'
import { useReducer, useCallback } from 'react'

import { MinDate, SavingGoalState, Actions } from 'app/models/SavingGoal'

const currencyMask = (money: number) => {
  return new Intl.NumberFormat('en-US', {
    style: 'currency',
    currency: 'USD',
    minimumFractionDigits: 2
  }).format(money)
}

const reducer = (state: SavingGoalState, action: Actions) => {
  switch (action.type) {
    case 'SET_TOTAL_AMOUNT':
      return { ...state, totalAmount: action.money }
    case 'SET_DATE':
      return { ...state, reachDate: action.date }
    case 'SET_MONTHLY_AMOUNT':
      return {
        ...state,
        monthlyAmount: action.amount
      }
  }
}

const useSavingGoal = ({ initialDate }: MinDate) => {
  const [state, dispatch] = useReducer(reducer, {
    reachDate: initialDate,
    totalAmount: '',
    monthlyAmount: ''
  })

  const setAmount = (money: string) => {
    dispatch({ type: 'SET_TOTAL_AMOUNT', money })
  }

  const setDateToReachGoal = (date: Date) => {
    dispatch({ type: 'SET_DATE', date })
  }

  const calcMonthlyAmount = useCallback(
    (dateDiff: number, totalAmount: string) => {
      const initialAmount = totalAmount === '0' ? '$0,00' : `${totalAmount}`

      if (dateDiff === 0) {
        dispatch({ type: 'SET_MONTHLY_AMOUNT', amount: initialAmount })
        return
      }

      const numberTotalAmount = Number(totalAmount.replace(/[,]+/g, ''))

      dispatch({
        type: 'SET_MONTHLY_AMOUNT',
        amount: currencyMask(numberTotalAmount / dateDiff)
      })
    },
    []
  )

  return {
    state,
    effects: {
      setAmount,
      setDateToReachGoal,
      calcMonthlyAmount
    }
  }
}

const [SavingGoalProvider, useStateSavingGoal, useEffectsSavingGoal] = constate(
  useSavingGoal,
  (value) => value.state,
  (value) => value.effects
)

export { SavingGoalProvider, useStateSavingGoal, useEffectsSavingGoal }
