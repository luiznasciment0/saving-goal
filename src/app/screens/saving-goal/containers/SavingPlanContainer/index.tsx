import SavingPlan from '../../components/SavingPlan'

import { useEffect, useState } from 'react'
import differenceInMonths from 'date-fns/differenceInMonths'
import addMonths from 'date-fns/addMonths'

import {
  useStateSavingGoal,
  useEffectsSavingGoal
} from 'app/providers/SavingGoalProvider'

const initialDate = new Date(new Date().getFullYear(), new Date().getMonth())

const SavingPlanContainer = () => {
  const [dateDifference, setDateDifference] = useState(0)
  const { totalAmount, reachDate, monthlyAmount } = useStateSavingGoal()
  const { calcMonthlyAmount } = useEffectsSavingGoal()

  const month = reachDate.toLocaleString('en-US', {
    month: 'long'
  })

  const year = reachDate.getFullYear()

  useEffect(() => {
    if (totalAmount == 0) {
      setDateDifference(0)
    } else if (
      month ===
      addMonths(initialDate, 1).toLocaleString('en-US', {
        month: 'long'
      })
    ) {
      setDateDifference(1)
    } else {
      setDateDifference(differenceInMonths(reachDate, initialDate))
    }
    calcMonthlyAmount()
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [totalAmount, reachDate])

  return (
    <SavingPlan
      dateDifference={dateDifference}
      month={month}
      year={year}
      monthlyAmount={monthlyAmount}
      totalAmount={totalAmount}
    />
  )
}

export default SavingPlanContainer
