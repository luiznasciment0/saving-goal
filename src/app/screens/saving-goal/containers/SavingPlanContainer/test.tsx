import { render, screen } from '@testing-library/react'

import { SavingGoalProvider } from 'app/providers/SavingGoalProvider'
import SavingPlanContainer from '.'

describe('<SavingPlanContainer />', () => {
  it('should add a month when right arrow is pressed', () => {
    const { container } = render(
      <SavingGoalProvider initialDate={new Date(2021, 1)}>
        <SavingPlanContainer initialDate={new Date(2021, 1)} />
      </SavingGoalProvider>
    )

    const monthlyAmountLabel = screen.getByRole('heading', {
      name: /Monthly amount/i
    })
    expect(monthlyAmountLabel).toHaveTextContent('Monthly amount')

    const monthlyAmount = screen.getByLabelText(
      'Amount of money to be monthly saved'
    )
    expect(monthlyAmount).toHaveTextContent('$0.00')

    const monthlyAmountDetails = screen.getByLabelText('Monthly save details')
    expect(monthlyAmountDetails).toHaveTextContent(
      'Youâ€™re planning 1 monthly deposits to reach your $0.00 goal by February 2021.'
    )

    expect(container.firstChild).toMatchSnapshot()
  })
})
