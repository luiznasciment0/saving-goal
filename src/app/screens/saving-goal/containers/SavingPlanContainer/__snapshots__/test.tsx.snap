// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[
  `<SavingPlanContainer /> should add a month when right arrow is pressed 1`
] = `
.c0 {
  border: 1px solid #e9eef2;
  border-radius: 4px;
  margin-top: 2.4rem;
  margin-bottom: 3.2rem;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 2.7rem 2.4rem;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c4 {
  background-color: #f4f8fa;
  padding: 2.4rem 3.2rem;
  display: grid;
  place-items: center;
  text-align: center;
}

.c2 {
  font-size: 1.8rem;
  line-height: 2.16rem;
  font-family: 'Work Sans';
  color: #1e2a32;
  font-weight: 400;
}

.c3 {
  font-size: 2.4rem;
  line-height: 2.88rem;
  font-family: 'Rubik';
  color: #0079ff;
  font-weight: 500;
}

.c5 {
  font-size: 1.2rem;
  line-height: 1.6rem;
  font-family: 'Work Sans';
  color: #1c1e1f;
  font-weight: 400;
}

@media (min-width:560px) {
  .c0 {
    width: 50.9rem;
  }
}

@media (min-width:560px) {
  .c1 {
    padding-top: 3.1rem;
    padding-bottom: 2.3rem;
  }
}

@media (min-width:560px) {
  .c4 {
    text-align: left;
  }
}

@media (max-width:350px) {

}

<div
  class="c0"
>
  <div
    aria-label="Amount of money to be monthly saved"
    class="c1"
  >
    <h6
      class="c2"
      color="primaryBlack"
    >
      Monthly amount
    </h6>
    <p
      class="c3"
      color="secondaryBlue"
    >
      $0.00
    </p>
  </div>
  <div
    aria-label="Monthly save details"
    class="c4"
  >
    <p
      class="c5"
      color="secondaryBlack"
    >
      Youâ€™re planning 
      <strong>
        1
         monthly deposits
      </strong>
       
      to reach your
      <strong>
         $
        0.00
      </strong>
       goal by
      <strong>
         
        February
         
        2021
        .
      </strong>
    </p>
  </div>
</div>
`
